src/Numerics/ODESolvers/BackwardEulerSolvers.jl

```diff
@@ -128,9 +172,147 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
 end
 
 function (lin::LinBESolver)(Q, Qhat, α, p, t)
+    rhs! = EulerOperator(lin.f_imp!, -α)
+
     if lin.α != α
         @assert lin.isadjustable
         update_backward_Euler_solver!(lin, Q, α)
     end
-    linearsolve!(lin.factors, lin.solver, Q, Qhat, p, t)
+
+    if typeof(lin.solver) <: AbstractIterativeSystemSolver
+        FT = eltype(α)
+        preconditioner_update!(
+            rhs!,
+            rhs!.f!,
+            lin.preconditioner,
+            nothing,
+            FT(NaN),
+        )
+        linearsolve!(rhs!, lin.preconditioner, lin.solver, Q, Qhat, p, t)
+        preconditioner_counter_update!(lin.preconditioner)
+    else
+        linearsolve!(rhs!, lin.factors, lin.solver, Q, Qhat, p, t)
+    end
+end
+
+"""
+    struct NonLinearBackwardEulerSolver{NLS}
+        nlsolver::NLS
+        isadjustable::Bool
+        preconditioner_update_freq::Int64
+    end
+
+Helper type for specifying building a nonlinear backward Euler solver with a nonlinear
+solver.
+
+nlsolver: iterative nonlinear solver, i.e., JacobianFreeNewtonKrylovSolver
+isadjustable: TODO not used, might use for updating preconditioner
+preconditioner_update_freq:  relavent to Jacobian free -1: no preconditioner;
+                             positive number, update every freq times
+"""
+struct NonLinearBackwardEulerSolver{NLS}
+    nlsolver::NLS
+    isadjustable::Bool
+    # preconditioner_update_freq, -1: no preconditioner;
+    # positive number, update every freq times
+    preconditioner_update_freq::Int
+    function NonLinearBackwardEulerSolver(
+        nlsolver;
+        isadjustable = false,
+        preconditioner_update_freq = -1,
+    )
+        NLS = typeof(nlsolver)
+        return new{NLS}(nlsolver, isadjustable, preconditioner_update_freq)
+    end
+end
+
+
+"""
+    LinBESolver
+
+Concrete implementation of an `AbstractBackwardEulerSolver` to use nonlinear
+solvers of type `NLS`. See helper type
+[`NonLinearBackwardEulerSolver`](@ref)
+```
+    Q = Qhat + α f_imp(Q, param, time)
+```
+"""
+mutable struct NonLinBESolver{FT, F, NLS} <: AbstractBackwardEulerSolver
+    # Solve Q - α f_imp(Q) = Qrhs, not used
+    α::FT
+    # implcit operator
+    f_imp!::F
+    # jacobian action, which approximates drhs!/dQ⋅ΔQ , here rhs!(Q) = Q - α f_imp(Q)
+    jvp!::JacobianAction
+    # nonlinear solver
+    nlsolver::NLS
+    # whether adjust the time step or not, not used
+    isadjustable::Bool
+    # preconditioner, approximation of drhs!/dQ
+    preconditioner::AbstractPreconditioner
+
+end
+
+Δt_is_adjustable(nlsolver::NonLinBESolver) = nlsolver.isadjustable
+
+"""
+    setup_backward_Euler_solver(solver::NonLinearBackwardEulerSolver, Q, α, tendency!)
+
+Returns a concrete implementation of an `AbstractBackwardEulerSolver` that will
+solve for `Q` in nonlinear systems of the form of
+```
+    Q = Qhat + α f(Q, param, time)
+```
```

Comment:
backticks are redundant with indentantion
```suggestion

    Q = Qhat + α f(Q, param, time)

```